source/intersection.js

```diff
@@ -24,13 +23,20 @@ import uniq from './uniq';
  */
 var intersection = _curry2(function intersection(list1, list2) {
   var lookupList, filteredList;
+  var toKeep = new _Set();
+
   if (list1.length > list2.length) {
     lookupList = list1;
     filteredList = list2;
   } else {
     lookupList = list2;
     filteredList = list1;
   }
-  return uniq(_filter(flip(_includes)(lookupList), filteredList));
+
+  for (var i = 0; i < lookupList.length; i += 1) {
+    toKeep.add(lookupList[i]);
+  }
+
+  return uniq(_filter(toKeep.has.bind(toKeep), filteredList));
 });
 export default intersection;
```

Comment:
I think if we're adding this optimization, then we should consider undoing #1597, which has remained controversial since it was created.